presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainFilterUtils.java

```diff
@@ -152,6 +160,91 @@ public static TupleDomainFilter toFilter(Domain domain, int notInThreshold)
         return MultiRange.of(rangeFilters, nullAllowed, false);
     }
 
+    /**
+     * Returns the exclusive number list if it could use NOT IN filter optimization.
+     *
+     * We're not able to distinguish NOT IN statement from normal range compare
+     * E.g.
+     * NOT IN (9, 10, 25) would be [Long.MIN_VALUE, 8] [11, 24] [26, Long.MAX_VALUE]
+     * (num < 8 OR (num > 11 AND number < 24) OR num > 26) could be the same
+     *
+     * When count of exclusive numbers is under the threshold, we can use NOT IN optimization
+     * it would return exclusive number list; otherwise empty list.
+     */
+    private static Optional<List<Long>> getExclusives(List<BigintRange> bigintRanges, int notInThreshold)
+    {
+        if (bigintRanges.size() < 2) {
+            return Optional.empty();
+        }
+
+        long count = getExclusiveCount(bigintRanges);
+        if (count >= notInThreshold || count < 0) {
+            return Optional.empty();
+        }
+
+        ImmutableList.Builder<Long> exclusiveNumbers = new ImmutableList.Builder<>();
```

Comment:
Don't use the ImmutableList.Builder<Long>. You converted it back to array later anyways. Since we already know the size, you can just new an int[] here. It'll be more efficient.

Reply:
> Don't use the ImmutableList.Builder. You converted it back to array later anyways. Since we already know the size, you can just new an int[] here. It'll be more efficient.

Thanks! Good suggestion!